### PR TITLE
Update: add ignoreJSX option to no-extra-parens (Fixes #7444)

### LINKED ISSUE #7444: Proposal: Add JSX-specific overrides to `no-extra-parens`
In order to accommodate best practices established in the JSX world, the rule `no-extra-parens` needs to be modified to optionally ignore `JSX` components - the general standard is to wrap at least multiline JSX components in parens. However, this conflicts with the best practices in JavaScript in general where extra parens around expressions are discouraged.

**What rule do you want to change?**

[`no-extra-parens`](http://eslint.org/docs/rules/no-extra-parens)

**Does this change cause the rule to produce more or fewer warnings?**

Neither - the default setting should not cause more or fewer warnings to appear. However, this proposal will allow users to reduce some warnings when working with JSX components.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option: `ignoreJSX` with possible values:
- `all`: All JSX components are **excluded** from this rule
- `multiline`: All multiline JSX components are **excluded** from this rule
- `single-line`: All single-line JSX components are **excluded** from this rule

The default behaviour for this exception will be to to disallow extra parens around any `JSX` components, just as it is currently implemented.

**Please provide some example code that this change will affect:**

``` js
/* eslint no-extra-parens: ["error", "all", { "ignoreJSX": "all" }] */
// Ok - JSX components are excluded from this rule
const Component = (<div />)
const Component = <div />
const Component = (
  <div>
    <p />
  </div>
)
const Component = 
  <div>
    <p />
  </div>

/* eslint no-extra-parens: ["error", "all", { "ignoreJSX": "multiline" }] */
// Ok (extra parens not allowed)
const Component = <div />
// Ok (multiline JSX components are excluded from this rule)
const Component = (
  <div>
    <p />
  </div>
)
const Component = 
  <div>
    <p />
  </div>
// Warns (extra parens not allowed)
const Component = (<div />)

/* eslint no-extra-parens: ["error", "all", { "ignoreJSX": "single-line" }] */
// Ok (single-line JSX components are excluded from this rule)
const Component = <div />
const Component = (<div />)
// Ok (extra parens not allowed)
const Component = 
  <div>
    <p />
  </div>
// Warns (extra parens not allowed)
const Component = (
  <div>
    <p />
  </div>
)
```

**What does the rule currently do for this code?**

The rule currently flags all cases where a JSX component is wrapped in parens. The general best practice, however, is to wrap at least multiline JSX components in parens, for better readability.

**What will the rule do after it's changed?**

The rule will allow developers to **optionally** turn this rule off for `JSX` components (single-line, multiline or all). The default setting will be to treat parens around `JSX` components just like any other expression - it will warn if `JSX` components are wrapped in parens.

See also original discussion in #7350.

_/cc @platinumazure @ljharb_

### PR DESCRIPTION
**Pull request by Alaneor**
**What is the purpose of this pull request? (put an "X" next to item)**

- [ ] Documentation update
- [ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
- [ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
- [x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
- [ ] Add autofixing to a rule
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

**What changes did you make? (Give an overview)**

- New option added to `no-extra-parens` rule: `ignoreJSX`, with options available:
  - `"all"`: All JSX components are ignored
  - `"multiline"`: Multiline JSX components are ignored
  - `"single-line"`: Single line JSX components are ignored

**Is there anything you'd like reviewers to focus on?**

Please review everything thoroughly, the rule was already quite complex and I am new to working with rules in ESLint.

### CODE REVIEW

File: docs/rules/no-extra-parens.md [hunk 1]
```diff
@@ -23,6 +23,7 @@ This rule has an object option for exceptions to the `"all"` option:
 * `"conditionalAssign": false` allows extra parentheses around assignments in conditional test expressions
 * `"returnAssign": false` allows extra parentheses around assignments in `return` statements
 * `"nestedBinaryExpressions": false` allows extra parentheses in nested binary expressions
+* `"ignoreJSX": "all|multiline|single-line"` allows extra parentheses around all/multiline/single-line JSX components
```
Comment (ljharb) [inline docs/rules/no-extra-parens.md]: Which of these three choices represents the default? It seems like we'd want a "none" option - that way I wouldn't be forced to delete the key (I like documenting all possible options even if I don't use them). Could we add a "never"?

Comment (Alaneor) [inline docs/rules/no-extra-parens.md]: Ah, I actually had a `false` options for this field before, but removed it as it seemed the preferred way of disabling the option was to omit the key.

`"none"` sounds like a good default. I will add it.

Comment (Alaneor) [inline docs/rules/no-extra-parens.md]: Added `"none"` as default option.

File: docs/rules/no-extra-parens.md [hunk 2]
```diff
@@ -23,6 +23,7 @@ This rule has an object option for exceptions to the `"all"` option:
 * `"conditionalAssign": false` allows extra parentheses around assignments in conditional test expressions
 * `"returnAssign": false` allows extra parentheses around assignments in `return` statements
 * `"nestedBinaryExpressions": false` allows extra parentheses in nested binary expressions
+* `"ignoreJSX": "none|all|multiline|single-line"` allows extra parentheses around all/multiline/single-line JSX components. Defaults to `none`.
```
Comment (not-an-aardvark) [inline docs/rules/no-extra-parens.md]: I think the option name should be `multi-line` for consistency with `single-line`.

File: tests/lib/rules/no-extra-parens.js [hunk 3]
```diff
@@ -291,7 +314,61 @@ ruleTester.run("no-extra-parens", rule, {
         { code: "foo instanceof (bar instanceof baz)", options: ["all", { nestedBinaryExpressions: false }] },
         { code: "foo in (bar in baz)", options: ["all", { nestedBinaryExpressions: false }] },
         { code: "foo + (bar + baz)", options: ["all", { nestedBinaryExpressions: false }] },
-        { code: "foo && (bar && baz)", options: ["all", { nestedBinaryExpressions: false }] }
+        { code: "foo && (bar && baz)", options: ["all", { nestedBinaryExpressions: false }] },
+
+        // ignoreJSX === "all"
+        jsx("const Component = (<div />)", ["all", { ignoreJSX: "all" }]),
```
Comment (not-an-aardvark) [inline tests/lib/rules/no-extra-parens.js]: In my opinion, this call signature is a bit confusing -- I had no idea what was going on until I looked at the function signature above.

I would recommend either:

* Use defaults in the `RuleTester` constructor
```js
const ruleTester = new RuleTester({
    parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } }
});
```

* Make the `jsx` function accept an object, instead of two parameters. This would make the jsx test cases look similar to everything else:

```js
    {
        code: "foo",
        options: ["bar"]
    },
    jsx({
        code: "foo",
        options: ["baz"]
    }),
```

File: lib/rules/no-extra-parens.js [hunk 4]
```diff
@@ -73,6 +75,38 @@ module.exports = {
...
          */
         function ruleApplies(node) {
+            if (node.type === "JSXElement") {
+                const isSelfClosing = node.openingElement.selfClosing;
+
+                // Self-closing components do not have `closingElement`, so we must check the start
+                // and end lines of this particular component directly
+                const isSingleLine = isSelfClosing
+                    ? node.loc.start.line === node.loc.end.line
+                    : astUtils.isTokenOnSameLine(node.openingElement, node.closingElement);
```
Comment (not-an-aardvark) [inline lib/rules/no-extra-parens.js]: Is there a reason to use `astUtils.isTokenOnSameLine` here rather than `node.loc.start.line === node.loc.end.line` in both cases? In other words, could you use `node.loc.start.line === node.loc.end.line` even when the element is not self-closing?

File: docs/rules/no-extra-parens.md (hunk 2 above)
Comment (Alaneor) [inline docs/rules/no-extra-parens.md]: Renamed.

File: tests/lib/rules/no-extra-parens.js (hunk 3 above)
Comment (Alaneor) [inline tests/lib/rules/no-extra-parens.js]: Thanks for the suggestion. Using the defaults for `RuleTester` sound like the best way to do this. Did not know this is possible. Updated.

File: lib/rules/no-extra-parens.js (hunk 4 above)
Comment (Alaneor) [inline lib/rules/no-extra-parens.js]: Ah, of course! I spent quite some time figuring out how to detect multi-line, multi-element JSX structures, and only much later realised that a single-element, multi-line JSX element is possible. I guess I got "thought-boxed". Updated to only use the `loc` info of the current node.

**Comment by Alaneor** [comment]
@not-an-aardvark Thanks for your comments. I have remedied all of the issues you brought up. Thanks!

File: tests/lib/rules/no-extra-parens.js [hunk 5]
```diff
@@ -706,6 +767,43 @@ ruleTester.run("no-extra-parens", rule, {
...
+        ].join("\n"), "const Component = <div>\n  <p />\n</div>;", "JSXElement", 1, {
+            options: ["all", { ignoreJSX: "single-line" }]
+        }),
+        invalid([
+            "const Component = (<div",
+            "  prop={true}",
+            "/>)"
+        ].join("\n"), "const Component = <div\n  prop={true}\n/>", "JSXElement", 1, {
+            options: ["all", { ignoreJSX: "single-line" }]
+        })
```
Comment (not-an-aardvark) [inline tests/lib/rules/no-extra-parens.js]: Could you add a test case for `{ ignoreJSX: "none" }`? It would be good to make sure that this behaves as expected.

Comment (Alaneor) [inline tests/lib/rules/no-extra-parens.js]: Good idea - added.

File: docs/rules/no-extra-parens.md [hunk 6]
```diff
@@ -23,6 +23,7 @@ This rule has an object option for exceptions to the `"all"` option:
 * `"conditionalAssign": false` allows extra parentheses around assignments in conditional test expressions
 * `"returnAssign": false` allows extra parentheses around assignments in `return` statements
 * `"nestedBinaryExpressions": false` allows extra parentheses in nested binary expressions
+* `"ignoreJSX": "none|all|multi-line|single-line"` allows extra parentheses around all/multi-line/single-line JSX components. Defaults to `none`.
```
Comment (platinumazure) [inline docs/rules/no-extra-parens.md]: Wondering if the documentation here should say something like "no/all/multi-line/single-line" to be consistent with the available options?

File: docs/rules/no-extra-parens.md [hunk 7]
```diff
@@ -104,6 +105,68 @@ x = a + (b * c);
...
+    </div>
+)
+const Component = (
+    <div
+        prop={true}
+    />
+)
+```
+
+Examples of **incorrect** code for this rule with the `all` and `{ "ignoreJSX": "multi-line" }` options:
```
Comment (platinumazure) [inline docs/rules/no-extra-parens.md]: Conventionally, we've tended to put incorrect code examples before correct code examples (per configuration that is being demonstrated). Would you mind doing the same here?

File: docs/rules/no-extra-parens.md [hunk 8]
```diff
@@ -104,6 +105,68 @@ x = a + (b * c);
...
+
+Examples of **correct** code for this rule with the `all` and `{ "ignoreJSX": "single-line" }` options:
+
+```js
+/* eslint no-extra-parens: ["error", "all", { ignoreJSX: "single-line" }] */
+const Component = (<div />)
+const Component = (<div><p /></div>)
+```
+
+Examples of **incorrect** code for this rule with the `all` and `{ "ignoreJSX": "single-line" }` options:
```
Comment (platinumazure) [inline docs/rules/no-extra-parens.md]: As noted above, we usually put incorrect code before correct code for a given example.

File: lib/rules/no-extra-parens.js [hunk 9]
```diff
@@ -73,6 +75,32 @@ module.exports = {
...
+                const isSingleLine = node.loc.start.line === node.loc.end.line;
+
+                switch (IGNORE_JSX) {
+
+                    // Exclude this JSX element from linting
+                    case "all":
+                        return false;
+
+                    // Exclude this JSX element if it is multi-line element
+                    case "multi-line": {
```
Comment (platinumazure) [inline lib/rules/no-extra-parens.js]: Why do you have a block statement here, but not anywhere else in the switch?

Comment (Alaneor) [inline lib/rules/no-extra-parens.js]: Oops, forgot to remove it after a bit of refactoring. I had a `const` inside before.

File: docs/rules/no-extra-parens.md (hunk 8 above)
Comment (Alaneor) [inline docs/rules/no-extra-parens.md]: Moved up.

File: docs/rules/no-extra-parens.md (hunk 7 above)
Comment (Alaneor) [inline docs/rules/no-extra-parens.md]: Moved up.

File: docs/rules/no-extra-parens.md (hunk 6 above)
Comment (Alaneor) [inline docs/rules/no-extra-parens.md]: Updated, thanks!

**Comment by Alaneor** [comment]
@platinumazure Thanks for your feedback. I have incorporated your suggestions.

File: tests/lib/rules/no-extra-parens.js [hunk 10]
```diff
@@ -291,7 +298,61 @@ ruleTester.run("no-extra-parens", rule, {
         { code: "foo instanceof (bar instanceof baz)", options: ["all", { nestedBinaryExpressions: false }] },
         { code: "foo in (bar in baz)", options: ["all", { nestedBinaryExpressions: false }] },
         { code: "foo + (bar + baz)", options: ["all", { nestedBinaryExpressions: false }] },
-        { code: "foo && (bar && baz)", options: ["all", { nestedBinaryExpressions: false }] }
+        { code: "foo && (bar && baz)", options: ["all", { nestedBinaryExpressions: false }] },
+
+        // ignoreJSX === "all"
```
Comment (vitorbal) [inline tests/lib/rules/no-extra-parens.js]: Minor convention nitpick: could you change these comments so that they look like this:

```js
// ["all", {ignoreJSX: "all"}]
```
That seems to be the way other "divider" comments in this file do it (See [line 213](https://github.com/eslint/eslint/pull/7926/files#diff-654d0ea8cd0d760e74e00806df5e3125R213), for example).

Comment (Alaneor) [inline tests/lib/rules/no-extra-parens.js]: Updated, thanks for the suggestion!

**Comment by Alaneor** [comment]
Incorporated @vitorbal 's suggestions and rebased on current master. Thanks for your feedback, looking forward to having this merged!
